editor/src/clj/editor/resource_update.clj

```diff
@@ -61,33 +66,95 @@
 (defn- merge-resource-change-plans [& plans]
   (apply merge-with into plans))
 
-(defn- replace-resources-plan [{:keys [resource->old-node]} resources]
-  ;; creates new resource nodes for all resources, transfers overrides and outgoing arcs
-  ;; from old (if any) and deletes them.
-  {:new (vec resources)
-   :transfer-overrides (into [] (keep (fn [resource] (when-let [old-node (resource->old-node resource)] [resource old-node]))) resources)
-   :transfer-outgoing-arcs (into [] (keep (fn [resource] (when-let [old-node (resource->old-node resource)] [resource (gu/explicit-outputs old-node)]))) resources)
-   :delete (into [] (keep resource->old-node) resources)
-   })
+(defn- keep-existing-node?
+  "Check if we can safely keep the existing node in the graph. Every time a file
+  has an updated timestamp it will be reported as changed, but if the contents
+  are unchanged we don't need to reload the node."
+  [old-node new-resource evaluation-context]
+  (let [resource-type (resource/resource-type new-resource)]
+    ;; Determine if the file contents have changed by comparing the sha256
+    ;; value of the in-memory state of the old-node to the calculated sha256 of
+    ;; the on-disk state of the new-resource.
+    (if (:stateless? resource-type)
+      ;; Stateless resources do not have an in-memory state, so their sha256
+      ;; output will return the sha256 of their on-disk state. Thus, we cannot
+      ;; determine if the file has changed compared to their in-memory state,
+      ;; and must assume it is unsafe to keep the existing node in the graph.
+      false
+
+      ;; This is a stateful resource. Compare the sha256 of the in-memory state
+      ;; to the sha256 of the on-disk state of the new-resource.
+      (let [read-fn (:read-fn resource-type)
+            write-fn (:write-fn resource-type)
+            old-sha256 (try
+                         (g/node-value old-node :sha256 evaluation-context)
+                         (catch Exception _
+                           nil))
+            new-sha256 (try
+                         (if (and read-fn write-fn)
+                           (let [source-value (read-fn new-resource)
+                                 sanitized-content (write-fn source-value)]
+                             (DigestUtils/sha256Hex ^String sanitized-content))
+                           (with-open [input-stream (io/input-stream new-resource)]
+                             (DigestUtils/sha256Hex ^InputStream input-stream)))
+                         (catch Exception _
+                           nil))]
```

Comment:
Is it safe to assume that for every unchanged resource node its `:sha265` output is `(sha (write (read resource)))`?